adam/experiment/observer.py

```diff
@@ -85,6 +88,160 @@ def report(self) -> None:
+        )
+        self.outfile.write(
+            """
+            <script>
+            function myFunction(id) {
+              var x = document.getElementById(id);
+              if (x.style.display === "none") {
+                x.style.display = "block";
+              } else {
+                x.style.display = "none";
+              }
+            }
+            </script>
+            """
+        )
+
+    def log(
+        self,
+        *,
+        instance_number: int,
+        observer_name: str,
+        true_description: str,
+        learner_description: str,
+        situation_string: str,
+        perception_string: str,
+    ):
+
+        clickable_perception_string = f"""
+            <button onclick="myFunction('myPerception{instance_number}')">View Perception</button>
+            <div id="myPerception{instance_number}" style="display: none">
+            {perception_string}
+            </div>
+            """
+
+        self.outfile.write(
+            f"\n\t<table>\n"
+            f"\t\t<thead>\n"
+            f"\t\t\t<tr>\n"
+            f'\t\t\t\t<th colspan="3">\n'
+            f"\t\t\t\t\t<h2>Learning Instance: {observer_name} report number {instance_number}</h2>\n"
+            f"\t\t\t\t</th>\n\t\t\t</tr>\n"
+            f"\t\t</thead>\n"
+            f"\t\t<tbody>\n"
+            f"\t\t\t<tr>\n"
+            f"\t\t\t\t<td>\n"
+            f'\t\t\t\t\t<h3 id="situation-{instance_number}">Situation</h3>\n'
+            f"\t\t\t\t</td>\n"
+            f"\t\t\t\t<td>\n"
+            f'\t\t\t\t\t<h3 id="true-{instance_number}">True Description</h3>\n'
+            f"\t\t\t\t</td>\n"
+            f"\t\t\t\t<td>\n"
+            f'\t\t\t\t\t<h3 id="learner-{instance_number}">Learner\'s Description</h3>\n'
+            f"\t\t\t\t</td>\n"
+            f"\t\t\t\t<td>\n"
+            f'\t\t\t\t\t<h3 id="perception-{instance_number}">Learner Perception</h3>\n'
+            f"\t\t\t\t</td>\n"
+            f"\t\t\t</tr>\n"
+            f"\t\t\t<tr>\n"
+            f'\t\t\t\t<td valign="top">{situation_string}\n\t\t\t\t</td>\n'
+            f'\t\t\t\t<td valign="top">{true_description}</td>\n'
+            f'\t\t\t\t<td valign="top">{learner_description}</td>\n'
+            f'\t\t\t\t<td valign="top">{clickable_perception_string}\n\t\t\t\t</td>\n'
+            f"\t\t\t</tr>\n\t\t</tbody>\n\t</table>"
+        )
+        self.outfile.write("\n</body>")
+
+
+@attrs(slots=True)
+class HTMLLoggerObserver(
+    DescriptionObserver[SituationT, LinguisticDescriptionT, PerceptionT]
+):
+    r"""
+    Logs the true description and learner's descriptions throughout the learning process.
+    """
+    name: str = attrib(validator=instance_of(str))
+    html_logger: HTMLLogger = attrib(
+        init=True, validator=instance_of(HTMLLogger), kw_only=True
+    )
+    counter: int = attrib(kw_only=True, default=0)
+
+    def observe(  # pylint:disable=unused-argument
```

Comment:
It's a little cleaner if `observe` here does no work within this class but just forwards to a method on `HTMLLogger`.